flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/Timestamp.java

```diff
@@ -0,0 +1,202 @@
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+
+import static org.apache.calcite.avatica.util.DateTimeUtils.MILLIS_PER_DAY;
+import static org.apache.calcite.avatica.util.DateTimeUtils.MILLIS_PER_HOUR;
+import static org.apache.calcite.avatica.util.DateTimeUtils.MILLIS_PER_MINUTE;
+import static org.apache.calcite.avatica.util.DateTimeUtils.MILLIS_PER_SECOND;
+import static org.apache.calcite.avatica.util.DateTimeUtils.unixTimestamp;
+
+/**
+ * Immutable SQL TIMESTAMP type with nanosecond precision.
+ *
+ * <p>This class is composite of a millisecond and nanoOfMillisecond. The millisecond part
+ * holds the integral second and the milli-of-second. The nanoOfMillisecond holds the
+ * nano-of-millisecond, which should between 0 - 999_999.
+ */
+public class Timestamp implements Comparable<Timestamp> {
+
+	// nanos per second
+	public static final long NANOS_PER_SECOND = 1000_000_000L;
+
+	// nanos per minute
+	public static final long NANOS_PER_MINUTE = NANOS_PER_SECOND * 60;
+
+	// nanos per second
+	public static final long NANOS_PER_HOUR = NANOS_PER_MINUTE * 60;
+
+	// this field holds the integral second and the milli-of-second
+	private final long millisecond;
+
+	// this field holds the nano-of-millisecond
+	private final int nanoOfMillisecond;
+
+	public Timestamp(long millisecond, int nanoOfMillisecond) {
+		Preconditions.checkArgument(nanoOfMillisecond >= 0 && nanoOfMillisecond <= 999_999);
+		this.millisecond = millisecond;
+		this.nanoOfMillisecond = nanoOfMillisecond;
+	}
+
+	@Override
+	public int compareTo(Timestamp that) {
+		int cmp = Long.compare(this.millisecond, that.millisecond);
+		if (cmp == 0) {
+			cmp = this.nanoOfMillisecond - that.nanoOfMillisecond;
+		}
+		return cmp;
+	}
+
+	@Override
+	public boolean equals(Object obj) {
+		if (!(obj instanceof Timestamp)) {
+			return false;
+		}
+		Timestamp that = (Timestamp) obj;
+		return this.millisecond == that.millisecond &&
+				this.nanoOfMillisecond == that.nanoOfMillisecond;
+	}
+
+	@Override
+	public String toString() {
+		return toLocalDateTime().toString();
+	}
+
+	@Override
+	public int hashCode() {
+		int ret = (int) millisecond ^ (int) (millisecond >> 32);
+		return 31 * ret + nanoOfMillisecond;
+	}
+
+	public long getMillisecond() {
+		return millisecond;
+	}
+
+	public int getNanoOfMillisecond() {
+		return nanoOfMillisecond;
+	}
+
+	/**
+	 * Obtains an instance of {@code Timestamp} from a millisecond.
+	 *
+	 * <p>This returns a {@code Timestamp} with the specified millisecond.
+	 * The nanoOfMillisecond field will be set to zero.
+	 *
+	 * @param millisecond the number of milliseconds since January 1, 1970, 00:00:00 GMT
+	 *                    A negative number is the number of milliseconds before
+	 *                    January 1, 1970, 00:00:00 GMT
+	 * @param precision the precision of fractional seconds
+	 * @return an instance of {@code Timestamp}
+	 */
+	public static Timestamp fromLong(long millisecond, int precision) {
```

Comment:
Isn't it that only compact precision can come from long?
This `millisecond` is different from `Timestamp.millisecond`?

Reply:
Yes, `Timestamp.fromLong(123L, 0)` obtains a `Timestamp` instance with `Timestamp.millisecond` == 0. Since the precision is 0, the milli part would be dropped.